programs/perpetuals/src/state/custody.rs

```diff
@@ -289,6 +358,121 @@ impl Custody {
 
         Ok(())
     }
+
+    pub fn get_collective_position(&self, side: Side) -> Result<Position> {
+        let stats = if side == Side::Long {
+            &self.long_positions
+        } else {
+            &self.short_positions
+        };
+        if stats.open_positions > 0 {
+            Ok(Position {
+                side,
+                price: math::checked_as_u64(math::checked_div(
+                    stats.weighted_leverage,
+                    stats.total_leverage,
+                )?)?,
+                size_usd: stats.size_usd,
+                collateral_usd: stats.collateral_usd,
+                unrealized_loss_usd: stats.cumulative_interest,
+                cumulative_interest_snapshot: stats.cumulative_interest_snapshot,
+                locked_amount: stats.locked_amount,
+                ..Position::default()
+            })
+        } else {
+            Ok(Position::default())
+        }
+    }
+
+    pub fn add_position(&mut self, position: &Position, curtime: i64) -> Result<()> {
+        // compute accumulated interest
+        let collective_position = self.get_collective_position(position.side)?;
+        let interest = self.get_interest_amount_usd(&collective_position, curtime)?;
+
+        // update positions
+        let stats = if position.side == Side::Long {
+            &mut self.long_positions
+        } else {
+            &mut self.short_positions
+        };
+
+        stats.cumulative_interest = math::checked_add(stats.cumulative_interest, interest)?;
+        stats.cumulative_interest_snapshot = position.cumulative_interest_snapshot;
```

Comment:
it will be same because `position.cumulative_interest_snapshot` is getting from `stats.cumulative_interest_snapshot`

Reply:
right, `position.cumulative_interest_snapshot` is calculated as `custody.get_cumulative_interest(curtime)?` and this is what `stats.cumulative_interest_snapshot` should be set to upon the update